File: 2021/10/28/matrix_cells_distance_order.py
"""
https://leetcode.com/problems/matrix-cells-in-distance-order/
Author: Andrew Jarombek
Date: 10/28/2021
"""

from typing import List


def all_cells_dist_order(rows: int, cols: int, r_center: int, c_center: int) -> List[List[int]]:
    result = [[r_center, c_center]]
    result_set = set()
    result_set.add((r_center, c_center))
    index = 0

    while index < len(result):
        r = result[index][0]
        c = result[index][1]

        if r > 0 and (r - 1, c) not in result_set:
            result.append([r - 1, c])
            result_set.add((r - 1, c))

        if r < rows - 1 and (r + 1, c) not in result_set:
            result.append([r + 1, c])
            result_set.add((r + 1, c))

        if c > 0 and (r, c - 1) not in result_set:
            result.append([r, c - 1])
            result_set.add((r, c - 1))

        if c < cols - 1 and (r, c + 1) not in result_set:
            result.append([r, c + 1])
            result_set.add((r, c + 1))

        index += 1

    return result

File: 2022/06/07/matrixCellsDistanceOrder.ts
/**
 * https://leetcode.com/problems/matrix-cells-in-distance-order/
 * @author Andrew Jarombek
 * @since 6/7/2022
 */

function allCellsDistOrder(rows: number, cols: number, rCenter: number, cCenter: number): number[][] {
    let result = [];

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            result.push([i, j]);
        }
    }

    result.sort((a, b) => {
        return (Math.abs(a[0] - rCenter) + Math.abs(a[1] - cCenter)) -
            (Math.abs(b[0] - rCenter) + Math.abs(b[1] - cCenter))
    });

    return result;
}